add binary data handling to `bits` commands

astral-l opened the pull request:
<!--
if this PR closes one or more issues, you can automatically link the PR with
them by using one of the [*linking keywords*](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword), e.g.
- this PR should close #xxxx
- fixes #xxxx

you can also mention related issues, PRs or discussions!
-->

# Description
<!--
Thank you for improving Nushell. Please, check our [contributing guide](../CONTRIBUTING.md) and talk to the core team before making major changes.

Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.
-->
- enables `bits` commands to operate on binary data, where both inputs are binary and can vary in length
- adds an `--endian` flag to `bits and`, `or`, `xor` for specifying endianness (for binary values of different lengths)

# User-Facing Changes
<!-- List of all changes that impact the user experience here. This helps us keep track of breaking changes. -->
- `bits` commands will no longer error for non-int inputs

# Tests + Formatting
<!--
Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
- `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```
-->

> addendum: first PR, please inform if any changes are needed

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/and.rs:
```diff
@@ -17,17 +30,32 @@ impl Command for BitsAnd {
 
     fn usage(&self) -> &str {
-        "Performs bitwise and for ints."
+        "Performs bitwise and for ints or binary."
```
Nitpick: "ints or binary values" might be more consistent

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/and.rs:
```diff
@@ -42,51 +70,131 @@ impl Command for BitsAnd {
         vec![
             Example {
-                description: "Apply bits and to two numbers",
+                description: "Apply logical and to two numbers",
```
Correct me if I'm wrong, but I think these are all bitwise ands, not logical ands? I'm not sure why the existing example below said "Apply logical and to a list of numbers"

astral-l replied:
thanks for catching 👍

ysthakur commented:
Given that `or`, `and`, and `xor` share quite a bit of code, it may help to extract some of their shared functionality to a `util` file with a helper or two (to clarify, this is just something you can do if you think it's a good idea, I'm not saying it would necessarily help make the code significantly more readable/easier to maintain and that you should do it)

astral-l commented:
> Given that `or`, `and`, and `xor` share quite a bit of code, it may help to extract some of their shared functionality to a `util` file with a helper or two

just did this, I think it improves readability but let me know your thoughts as well

ysthakur commented:
Looks good, but I was actually thinking the call to `operate` itself could be replaced by a call to a helper, to reduce the code duplication even more. Would that be possible?

astral-l commented:
how about this? removing the call to `operate` means there is no need for an `Arguments` struct anymore for those commands, I am not sure if the same can be applied to the remainder of the commands since they are internally more different from each other.

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/not.rs:
```diff
@@ -119,14 +144,23 @@ impl Command for BitsNot {
+        number_size,
+    } = *args;
+    match input {
         Value::Int { val, .. } => {
+            let val = *val;
```
```suggestion
        Value::Int { &val, .. } => {
```
To avoid that extra assignment

astral-l replied:
for me this does not compile (parse error), and using `ref val` instead does not have the expected effect

ysthakur replied:
Ah sorry, never mind then

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/rotate_left.rs:
```diff
@@ -96,61 +123,82 @@ impl Command for BitsRol {
+                ),
             }
         }
-        // Propagate errors by explicitly matching them before the final case.
-        Value::Error { .. } => value,
```
I think you still need to propagate errors

ysthakur commented:
> how about this? removing the call to operate means there is no need for an Arguments struct anymore for those commands, I am not sure if the same can be applied to the remainder of the commands since they are internally more different from each other.

That looks pretty good. I think you can stuff the endian processing code into the helper too (it can return a `Value::error` if the given endianness is invalid).

About the `--number-bytes` flag, since it's currently being kept as only a string, maybe you could leave the `get_number_bytes` helper in `mod.rs` and the number-bytes-related code in all the commands as is for now? I'm not entirely sure why the `auto` option is necessary, since that's the default behavior and without it the flag could use `SyntaxShape::Int`, but that's a different discussion.

I'm not sure about the `usize`/`i64` thing either, gonna take a look later. I saw you asked about it in Discord already, you may want to ask again (later, after waiting a while) if no one answers.

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/mod.rs:
```diff
@@ -95,7 +100,68 @@ fn get_input_num_type(val: i64, signed: bool, number_size: NumberBytes) -> Input
+        (Value::Binary { val: lhs, .. }, Value::Binary { val: rhs, .. }) => {
+            let (lhs, rhs, max_len, min_len) = match (lhs.len(), rhs.len()) {
+                (max, min) if max > min => (lhs, rhs, max, min),
+                (min, max) => (rhs, lhs, min, max),
```
I'm getting a crash when the left operand is smaller:

![image](https://github.com/nushell/nushell/assets/45539777/db79a77b-1c82-47b8-afb6-eeb6a7a10d3e)

I think this should fix it? Make sure to add a test/example for that scenario first, though (I think an example might be better, so users have a better idea of `and`'s behavior)
```suggestion
                (min, max) => (rhs, lhs, max, min),
```
Also, sorry I didn't review this PR thoroughly earlier, been busy

astral-l replied:
right thanks, that was what I intended and not sure how I missed it
I am currently compiling the fix to verify it is correct, then I will add a few more tests as well

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/mod.rs:
```diff
@@ -95,7 +100,68 @@ fn get_input_num_type(val: i64, signed: bool, number_size: NumberBytes) -> Input
+            };
+
+            let bytes: Vec<u8> = lhs
+                .iter()
```
Does this handle the case when `lhs` is smaller than `rhs` and the user wants to use little endian? It feels like `lhs` should be padded too

astral-l commented:
just fixed, I also noticed that the endianness behavior was doing the opposite of what was expected:
```
$ 0x[ff] | bits or 0x[12 34] --endian big
0x[ff 34]
$ 0x[ff] | bits or 0x[12 34] --endian little
0x[12 ff]
```
which is confusing, so I swapped them around and updated all relevant examples

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/mod.rs:
```diff
@@ -95,7 +100,68 @@ fn get_input_num_type(val: i64, signed: bool, number_size: NumberBytes) -> Input
+            ShellError::OnlySupportsThisInputType {
+                exp_input_type: "int or binary".into(),
+                wrong_type: other.get_type().to_string(),
+                dst_span: other.span(),
+                src_span: span,
```
One of these spans needs to be `call.head`, I think. Here's the old error:
![image](https://github.com/nushell/nushell/assets/45539777/337c82bb-822a-4b0d-bd7f-fef412590705)

and the new one:

![image](https://github.com/nushell/nushell/assets/45539777/46dbfcad-89e9-4355-9891-695810281df7)

astral-l replied:
should be fixed now, spans are confusing
thanks again!

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/rotate_left.rs:
```diff
@@ -96,61 +111,93 @@ impl Command for BitsRol {
                 wrong_type: other.get_type().to_string(),
-                dst_span: head,
-                src_span: other.span(),
+                dst_span: other.span(),
+                src_span: span,
```
Make sure to swap these too

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/rotate_left.rs:
```diff
@@ -96,61 +111,93 @@ impl Command for BitsRol {
+            let bits = bits as u64;
+            let input_num_type = get_input_num_type(val, signed, number_size);
+            match input_num_type {
+                One if bits <= 0xff => {
```
I'm a bit unclear on these guards. Is rotating single byte integers more than 2^255 something that isn't useful/doesn't make sense?

astral-l replied:
the original implementation uses `i64::try_from`, and errors when trying to rotate an 8-byte int into the negative range (e.g. `1 | bits ror 1 -n 8`).

I think this was supposed to be checking if `val` (instead of `bits`) was in range for the specified int size so that the error could be moved up into the match statement, but that isn't what's being checked and shouldn't be necessary since it's already handled by `get_input_num_type`.
just fixed to put this back in line with the original behavior

fwiw, I am not sure it is necessary to error when rotating i64s into the signed range, but do not know what the implications of changing this default are

ysthakur replied:
I see. I'm not sure what the implications to not having it error are either, but changes to existing behavior should be avoided. Also, I do think errorring without `--signed` is reasonable, in case someone is dealing with naturals and doesn't want to get any negative numbers back. That said, you can ask about it in Discord and maybe see if people are fine with it first.

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/rotate_left.rs:
```diff
@@ -96,61 +111,93 @@ impl Command for BitsRol {
-                SignedFour => get_rotate_left(val as i32, bits, span),
-                SignedEight => get_rotate_left(val, bits, span),
+            let val = *val;
+            let bits = bits as u64;
```
Since `bits` is always converted to a `u32`, perhaps it could be converted to a `u32` right here? (and an error thrown if it's bigger than a `u32`?)

ysthakur commented:
There's also bounds checks in shl and shr that I think could be removed. Those two commands could be refactored the same way you did rol and ror.

Other than that, I think this PR looks pretty good! Thanks for continuing to work on this. After that one change, I think this'll be ready to merge

astral-l commented:
thanks for your patience and thorough review! I am currently working using a phone so it is not as easy to keep track of everything. hopefully it checks out now

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/and.rs:
```diff
@@ -42,52 +58,82 @@ impl Command for BitsAnd {
+            None => cfg!(target_endian = "little"),
+        };
 
-        // This doesn't match explicit nulls
```
I'm not sure if this comment can be removed

astral-l replied:
removed this when I changed the implementation of `and, or, xor` to use `operate` which made that section redundant, but since changing it back I forgot to add back the comment. fixing next commit

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/or.rs:
```diff
@@ -42,14 +59,37 @@ impl Command for BitsOr {
+            None => cfg!(target_endian = "little"),
+        };
 
-        // This doesn't match explicit nulls
```
This comment too (sorry, I'd make a suggestion to make it easier for you to just click the button and commit but GitHub doesn't let you do that on deleted lines)

ysthakur commented on crates/nu-cmd-extra/src/extra/bits/xor.rs:
```diff
@@ -42,13 +59,37 @@ impl Command for BitsXor {
+            }
+            None => cfg!(target_endian = "little"),
+        };
+
```
```suggestion

// This doesn't match explicit nulls
```

ysthakur reviewed (approved):
Great work @astral-l! (especially since you're working on your phone)

I'm not familiar enough with this part of the codebase to merge so I'm going to ask someone else to, but this looks good to me.

WindSoilder commented on crates/nu-cmd-extra/src/extra/bits/and.rs:
```diff
@@ -42,52 +58,82 @@ impl Command for BitsAnd {
         let head = call.head;
-        let target: i64 = call.req(engine_state, stack, 0)?;
+        let target: Value = call.req(engine_state, stack, 0)?;
+        let endian = call.get_flag::<Value>(engine_state, stack, "endian")?;
```
You can use `call.get_flag::<Spanned<String>>(engine_state, stack, "endian")`.  So you don't need to check `Value` variant

WindSoilder commented on crates/nu-cmd-extra/src/extra/bits/xor.rs:
```diff
@@ -42,13 +59,37 @@ impl Command for BitsXor {
-        let target: i64 = call.req(engine_state, stack, 0)?;
-        // This doesn't match explicit nulls
+        let target: Value = call.req(engine_state, stack, 0)?;
+        let endian = call.get_flag::<Value>(engine_state, stack, "endian")?;
```
The same here: `call.get_flag::<Spanned<String>>(engine_state, stack, "endian")`

WindSoilder commented on crates/nu-cmd-extra/src/extra/bits/or.rs:
```diff
@@ -42,14 +59,37 @@ impl Command for BitsOr {
         let head = call.head;
-        let target: i64 = call.req(engine_state, stack, 0)?;
+        let target: Value = call.req(engine_state, stack, 0)?;
+        let endian = call.get_flag::<Value>(engine_state, stack, "endian")?;
```
The same here

WindSoilder commented on crates/nu-cmd-extra/src/extra/bits/shift_right.rs:
```diff
@@ -56,29 +77,24 @@ impl Command for BitsShr {
     ) -> Result<PipelineData, ShellError> {
         let head = call.head;
-        let bits: usize = call.req(engine_state, stack, 0)?;
+        let bits: i64 = call.req(engine_state, stack, 0)?;
```
Can you tell me the reason to allow negative number input?  I saw you convert it to u32 on line 135 .

astral-l replied:
didn't know how the FromValue implementation for usize handles negative numbers, I just read it and see that it errors out which makes sense. will change this back to usize

WindSoilder commented on crates/nu-cmd-extra/src/extra/bits/shift_right.rs:
```diff
@@ -56,29 +77,24 @@ impl Command for BitsShr {
+        let bits: i64 = call.req(engine_state, stack, 0)?;
         let signed = call.has_flag(engine_state, stack, "signed")?;
-        let number_bytes: Option<Spanned<String>> =
+        let number_bytes: Option<Spanned<i64>> =
```
Not really sure what if you use `usize`?  Because it only support 1, 2, 4, 8, no negative number

astral-l replied:
agree with this as well

WindSoilder reviewed (approved):
Thank you!  LGTM now!
